services/src/pro/datasets/in_memory.rs

```diff
@@ -181,24 +201,29 @@ impl DatasetStore<UserSession> for ProHashMapDatasetDb {
 }
 
 #[async_trait]
-impl DatasetProvider for ProHashMapDatasetDb {
+impl DatasetProvider<UserSession> for ProHashMapDatasetDb {
     async fn list(
         &self,
-        // _session: &UserSession,
+        session: &UserSession,
         options: Validated<DatasetListOptions>,
     ) -> Result<Vec<DatasetListing>> {
-        // TODO: permissions
-
-        // TODO: include datasets from external dataset providers
         let options = options.user_input;
 
+        let iter = self
+            .dataset_permissions
+            .iter()
+            .filter(|p| session.roles.contains(&p.role))
+            .map(|p| {
+                self.datasets
+                    .get(&p.dataset)
+                    .expect("a dataset has at least one permission")
```

Comment:
It is rather: You wouldn't have got here if the filter wouldn't have matched and this means: There is a permission :smile: .
But you don't check that the dataset has a permissions but rather that the dataset exists for that permission.

Reply:
adab347397faa698db32ae904b15882856f6dc56

Reply:
Should we really panic here?

I guess it is one thing to say:
```
len() > 0
get(0).expect("should be there")
```

The other thing is
```
assert!(a[0].is_some());
b[0].expect("must be there, we have this assumption")
```

I guess the one thing cannot happen and the other thing means, we fucked up in consistency.
But would be panic then?

Reply:
instead of `expect` I could `.filter_map(|p| self.datasets.get(&p.dataset));` but then it would be hard to debug why a dataset is not listed. Shall I log a warning then?

Reply:
67fc660061e6cc71840b87653ed066137a5f9415